cdmtaskservice/s3/client.py

```diff
@@ -0,0 +1,207 @@
+        size - the total size of the object
+        part_size - the part size used in a multipart upload, if relevant.
+        """
+        # input checking seems a little pointless. Could check that the path looks like a path
+        # and the e-tag looks like an e-tag, but since this module should be creating this
+        # class why bother. This is why access modifiers for methods are good
+        self.path = path
+        self.e_tag = e_tag
+        self.size = size
+        self.part_size = part_size
+        # could check that the part size makes sense given the e-tag... meh
+    
+    @property
+    def has_parts(self) -> bool:
+        """ Returns true if the object was uploaded as multipart. """
+        return len(self.e_tag) > 32
+
+    @property
+    def num_parts(self) -> int:
+        """ 
+        Returns the number of parts used in a multipart upload or 1 if the upload was not
+        multipart.
+        """
+        if len(self.e_tag) > 32:
+            return int(self.e_tag.split("-")[1])
+        return 1
+
+
+class S3Client:
+    """
+    The S3 client.
+    
+    Note the client is *not threadsafe* as the underlying aiobotocore session is not threadsafe.
+    """
+    
+    @classmethod
+    async def create(
+        cls,
+        endpoint_url: str,
+        access_key: str,
+        secret_key: str,
+        config: dict[str, Any] = None,
+        skip_connection_check: bool = False
+    ) -> Self:
+        """
+        Create the client.
+        
+        endpoint_url - the URL of the s3 endpoint.
+        access_key - the s3 access key.
+        secret_key - the s3 secret key.
+        config - Any client configuration options provided as a dictionary. See
+            https://botocore.amazonaws.com/v1/documentation/api/latest/reference/config.html
+        skip_connetion_check - don't try to list_buckets while creating the client to check
+            the host and credentials are correct.
+        """
+        s3c = S3Client(endpoint_url, access_key, secret_key, config)
+        if not skip_connection_check:
+            async def list_buckets(client):
+                return await client.list_buckets()
+            await s3c._run_command(list_buckets)
+        return s3c
+    
+    def __init__(
+        self, endpoint_url: str, access_key: str, secret_key: str, config: dict[str, Any]
+    ):
+        self._url = self._require_string(endpoint_url, "endpoint_url")
+        self._ak = self._require_string(access_key, "access_key")
+        self._sk = self._require_string(secret_key, "secret_key")
+        self._config = Config(**config) if config else None
+        self._sess = get_session()
+    
+    def _require_string(self, string, name):
+        if not string or not string.strip():
+            raise ValueError(f"{name} is required")
+        return string.strip()
+    
+    def _client(self):
+        # Creating a client seems to be pretty cheap, usually < 20ms.
+        return self._sess.create_client(
```

Comment:
Is this equivalent to to `s3 = boto3.client('s3', ...)`? I believe using `boto3` is the standard method for interacting with an S3 client.

Reply:
That's my understanding. The client I'm using is an async version of boto3, and they both use the underlying botocore library

Reply:
👍